File: 0101-0200/0190.Reverse_Bits.py
#42ms




class Solution:
    def reverseBits(self, n: int) -> int:
        result = 0
        for _ in range(32):
            result = (result << 1) | (n & 1)
            n >>= 1
        return result
      





class Solution:
    def reverseBits(self, n: int) -> int:
        # Initialize result to 0 — this will store the reversed bit pattern
        result = 0
        
        # Loop exactly 32 times because we are working with 32-bit unsigned integers
        for _ in range(32):
            
            # (n & 1) extracts the least significant bit (LSB) of n
            # result << 1 shifts result left by 1 bit to make room for the new bit
            # OR operation combines them, adding the extracted bit to result
            result = (result << 1) | (n & 1)
            
            # Shift n right by 1 to process the next bit in the next iteration
            n >>= 1
        
        # Return the final reversed integer
        return result


'''

| Component             | Role                                                              |                                           |
| --------------------- | ----------------------------------------------------------------- | ----------------------------------------- |
| `result = 0`          | Stores the reversed bits as they are built                        |                                           |
| `for _ in range(32):` | Ensures processing of all 32 bits regardless of leading zeros     |                                           |
| `n & 1`               | Extracts the current least significant bit                        |                                           |
| `result << 1`         | Shifts previously stored bits left to make space for the next bit |                                           |
| `                     | `                                                                 | Inserts the extracted bit into the result |
| `n >>= 1`             | Moves to the next bit of the input number                         |                                           |
| `return result`       | Outputs the reversed 32-bit pattern                               |                                           |


'''


#38ms




class Solution:
    def reverseBits(self, n: int) -> int:
        return int(
            (
                bin(n)[2:].zfill(32)    # Convert n to binary string, drop the '0b' prefix, and pad with leading zeros to make it 32 bits
            )[::-1],                    # Reverse the 32-bit binary string
            2                            # Convert the reversed binary string back to an integer (base 2)
        )






'''



| Part          | Role                                                                          |
| ------------- | ----------------------------------------------------------------------------- |
| `bin(n)`      | Converts integer `n` to a binary string (e.g., `"0b101010"`).                 |
| `[2:]`        | Removes the `"0b"` prefix so only bits remain.                                |
| `.zfill(32)`  | Pads the string with leading zeros to ensure it has exactly 32 bits.          |
| `[::-1]`      | Reverses the 32-bit binary string.                                            |
| `int(..., 2)` | Converts the reversed binary string back to an integer using base-2 notation. |




'''





#37ms






class Solution:
    def reverseBits(self, n: int) -> int:
        x = 0
        for i in range(32):
            if (n >> i) & 1:
                x = x | (1 << (31 - i))
            else:
                x = x & ~(1 << (31 - i))
        return x






class Solution:
    def reverseBits(self, n: int) -> int:
        # x will store the final reversed 32-bit result
        x = 0
        
        # Loop through all 32 bit positions (i = 0 is LSB, i = 31 is MSB)
        for i in range(32):
            
            # Check whether the i-th bit of n is 1
            if (n >> i) & 1:
                # If the i-th bit of n is 1:
                # Set the (31 - i) th bit of x to 1
                x = x | (1 << (31 - i))
            else:
                # If the i-th bit of n is 0:
                # Clear the (31 - i) th bit of x (even though it is already 0 initially)
                x = x & ~(1 << (31 - i))
        
        # After all 32 bits are processed, return the reversed result
        return x



'''


| Component            | Role                                                     |
| -------------------- | -------------------------------------------------------- |
| `x = 0`              | Stores the output bits in reversed order                 |
| `for i in range(32)` | Iterates through every bit position of the 32-bit number |
| `(n >> i) & 1`       | Reads the bit at position `i` of `n`                     |
| `1 << (31 - i)`      | Computes the symmetric reversed bit position             |
| `x = x \| (...)`     | Sets a specific bit of `x` to `1`                        |
| `x = x & ~(...)`     | Clears a specific bit of `x` (forces it to `0`)          |
| `return x`           | Returns the final reversed 32-bit integer                |



'''






#37ms






class Solution:
    def reverseBits(self, n: int) -> int:
        val = format(n, "032b")
        val = val[::-1]
        op = int(val, 2)
        return op





class Solution:
    def reverseBits(self, n: int) -> int:
        # Convert n to a 32-bit binary string (padding with leading zeros if necessary)
        val = format(n, "032b")
        
        # Reverse the 32-bit binary string
        val = val[::-1]
        
        # Convert the reversed binary string back to an integer (base 2)
        op = int(val, 2)
        
        # Return the final reversed integer
        return op





'''

| Component           | Role                                                                                 |
| ------------------- | ------------------------------------------------------------------------------------ |
| `format(n, "032b")` | Converts integer `n` to a binary string of exactly 32 bits (zero-padded)             |
| `[::-1]`            | Reverses the binary string                                                           |
| `int(val, 2)`       | Converts the reversed binary string back into an integer using base-2 representation |
| `return op`         | Outputs the final reversed 32-bit integer                                            |




'''






#45ms






class Solution:
    def reverseBits(self, n: int) -> int:
        resultado = 0
        for _ in range(32):
            bit = n & 1
            n >>= 1
            resultado <<= 1
            resultado |= bit
        return resultado




class Solution:
    def reverseBits(self, n: int) -> int:
        # resultado will store the reversed bit pattern as it is constructed
        resultado = 0
        
        # Loop exactly 32 times because we are working with a 32-bit unsigned integer
        for _ in range(32):
            
            # Extract the least significant bit (LSB) from n
            bit = n & 1
            
            # Shift n to the right by 1 to move to the next bit for the next iteration
            n >>= 1
            
            # Shift resultado left by 1 to make space for the new bit being added
            resultado <<= 1
            
            # Add the extracted bit to resultado
            resultado |= bit
        
        # Return the final reversed integer after all 32 bits were processed
        return resultado




'''


| Component            | Role                                                          |                                                                              |
| -------------------- | ------------------------------------------------------------- | ---------------------------------------------------------------------------- |
| `resultado = 0`      | Stores the reversed bits of `n` as they are built             |                                                                              |
| `for _ in range(32)` | Ensures all 32 bits are processed even if leading zeros exist |                                                                              |
| `bit = n & 1`        | Extracts the current least significant bit from `n`           |                                                                              |
| `n >>= 1`            | Moves to the next bit of `n` by shifting right                |                                                                              |
| `resultado <<= 1`    | Shifts stored bits left to prepare for the next inserted bit  |                                                                              |
| `resultado           | = bit`                                                        | Inserts the extracted bit into the least significant position of `resultado` |
| `return resultado`   | Outputs the 32-bit reversed result                            |                                                                              |




'''

File: 0101-0200/0190.Reverse_Bits.rs
//2ms




impl Solution {
    pub fn reverse_bits(n: i32) -> i32 {
        n.reverse_bits()        
    }
}







// Defines a structure named 'Solution'. This is a standard convention 
// in many coding challenge platforms (like LeetCode) to organize code, 
// even though the struct itself has no fields or state here.
impl Solution {

    // Defines a public function named 'reverse_bits' within the Solution structure.
    // It accepts one argument, 'n', which is a 32-bit signed integer (i32),
    // and it is declared to return a 32-bit signed integer (i32).
    pub fn reverse_bits(n: i32) -> i32 {

        // This is the core line of logic.
        // Rust's primitive integer types (`i32`, `u32`, etc.) have a built-in 
        // method called `.reverse_bits()` provided by the standard library.
        // It efficiently performs the exact operation requested by the problem.
        // The return value of this method is automatically returned by the function.
        n.reverse_bits()        
    }

// Closes the implementation block.
}





//2ms






impl Solution {
    pub fn reverse_bits(n: i32) -> i32 {
        let (mut res, mut n) = (0u32, n as u32);
        for _ in 0..32 {
            res = (res << 1) | (n & 1);
            n >>= 1;
        }
        res as i32
    }
}







// This defines an 'implementation block' for a struct named 'Solution'. 
// This is a common pattern in coding challenges (like LeetCode) to organize methods.
impl Solution {

    // This declares a public function named 'reverse_bits'. 
    // It takes one argument 'n' (a 32-bit signed integer, i32) 
    // and promises to return a 32-bit signed integer (i32).
    pub fn reverse_bits(n: i32) -> i32 {

        // This line initializes two mutable variables using a tuple pattern:
        // 1. `res` (result): starts at 0, explicitly typed as `u32` (unsigned 32-bit integer).
        // 2. `n` (input copy): the original `i32` input is cast (`as u32`) to an unsigned 32-bit integer 
        //    to allow uniform bitwise operations inside the loop.
        let (mut res, mut n) = (0u32, n as u32);

        // This initiates a loop that runs exactly 32 times (from 0 up to, but not including, 32).
        // The `_` means we don't care about the loop index variable itself, just that it runs 32 times.
        for _ in 0..32 {

            // This is the core logic line:
            // 1. `(res << 1)`: Shifts the current result one position to the left, making space for the next bit.
            // 2. `(n & 1)`: Extracts the Least Significant Bit (LSB) of the input `n` (this will be 0 or 1).
            // 3. `|`: Uses the bitwise OR operator to place that extracted bit into the newly opened space in `res`.
            res = (res << 1) | (n & 1);

            // Shifts the input `n` one position to the right, discarding the bit we just processed 
            // and moving the *next* bit we need to the LSB position for the next loop iteration.
            n >>= 1;

        // The loop ends here. After 32 iterations, `res` holds all original bits in reverse order.
        }

        // Before returning, the final unsigned result (`u32`) is cast back 
        // to a signed 32-bit integer (`i32`) to match the function's return type signature.
        res as i32

    // The function definition ends here.
    }
// The implementation block ends here.
}








//0ms








impl Solution {
    pub fn reverse_bits(n: i32) -> i32 {
        // 1. Cast the input 'n' to a u32 type and make it mutable
        let mut n_u = n as u32; 

        // 2. Perform all operations on the unsigned integer 'n_u'
        // We also append 'u32' to the masks to explicitly define them as unsigned literals
        n_u = (n_u & 0b11111111111111110000000000000000u32) >> 16 | (n_u & 0b00000000000000001111111111111111u32) << 16;
        n_u = (n_u & 0b11111111000000001111111100000000u32) >> 8  | (n_u & 0b00000000111111110000000011111111u32) << 8;
        n_u = (n_u & 0b11110000111100001111000011110000u32) >> 4  | (n_u & 0b00001111000011110000111100001111u32) << 4;
        n_u = (n_u & 0b11001100110011001100110011001100u32) >> 2  | (n_u & 0b00110011001100110011001100110011u32) << 2;
        n_u = (n_u & 0b10101010101010101010101010101010u32) >> 1  | (n_u & 0b01010101010101010101010101010101u32) << 1;

        // 3. Cast the final result back to i32 for the return type
        return n_u as i32;
    }
}





//oms





impl Solution {
    pub fn reverse_bits(mut n: i32) -> i32 {
        let mut reversed = 0;
        for _ in 0..32 {
            let bit = n & 1;
            n >>= 1;
            reversed <<= 1;
            reversed = reversed | bit;
        }
        reversed
    }
}






// This defines an 'implementation block' for a struct named 'Solution'. 
// It's a standard pattern for organizing methods in coding challenges.
impl Solution {

    // This declares a public function named 'reverse_bits'. 
    // It takes one argument 'n', marked 'mut' (mutable) so we can change it inside the function.
    // It expects an 'i32' (signed 32-bit integer) and returns an 'i32'.
    pub fn reverse_bits(mut n: i32) -> i32 {

        // Initializes a mutable variable 'reversed' to store the result. 
        // It starts at 0, inferred as an i32 type.
        let mut reversed = 0;

        // Starts a loop that executes exactly 32 times (once for each bit in the 32-bit integer).
        for _ in 0..32 {

            // Extracts the Least Significant Bit (LSB) from 'n' using a bitwise AND operation with 1.
            // If the last bit of 'n' is 1, 'bit' becomes 1; otherwise, 'bit' becomes 0.
            let bit = n & 1;

            // Shifts 'n' one position to the right, effectively dropping the LSB we just captured 
            // and moving the next bit into the LSB position for the next iteration.
            // (Note: In Rust/most C-like languages, this is an arithmetic shift for i32, which might introduce issues with negative numbers).
            n >>= 1;

            // Shifts the current 'reversed' result one position to the left, making an empty space 
            // at the LSB position to insert the new bit we found.
            reversed <<= 1;

            // Inserts the extracted 'bit' into the LSB of the 'reversed' variable using 
            // the bitwise OR operation.
            reversed = reversed | bit;

        // The loop finishes here after 32 iterations.
        }
        
        // Returns the final accumulated value in the 'reversed' variable. 
        // Rust automatically uses the last expression as the return value if there is no semicolon.
        reversed
    
    // The function definition ends here.
    }
// The implementation block ends here.
}

